pkg/client/attach.go

```diff
@@ -170,36 +171,43 @@ func (c *ConmonClient) attach(ctx context.Context, cfg *AttachConfig) (err error
 		return nil
 	}
 
-	receiveStdoutError, stdinDone := c.setupStdioChannels(cfg, conn)
+	id := uuid.NewString()
+
+	receiveStdoutError, stdinDone := c.setupStdioChannels(cfg, conn, id)
 	if cfg.PostAttachFunc != nil {
 		if err := cfg.PostAttachFunc(); err != nil {
 			return fmt.Errorf("run post attach func: %w", err)
 		}
 	}
 
-	if err := c.readStdio(cfg, conn, receiveStdoutError, stdinDone); err != nil {
+	if err := c.readStdio(cfg, conn, id, receiveStdoutError, stdinDone); err != nil {
 		return fmt.Errorf("read stdio: %w", err)
 	}
 
 	return nil
 }
 
 func (c *ConmonClient) setupStdioChannels(
-	cfg *AttachConfig, conn *net.UnixConn,
+	cfg *AttachConfig, conn *net.UnixConn, id string,
 ) (receiveStdoutError, stdinDone chan error) {
 	receiveStdoutError = make(chan error)
 	go func() {
 		receiveStdoutError <- c.redirectResponseToOutputStreams(cfg, conn)
 	}()
 
 	stdinDone = make(chan error)
-	go func() {
-		var err error
-		if cfg.Streams.Stdin != nil {
-			_, err = util.CopyDetachable(conn, cfg.Streams.Stdin, cfg.DetachKeys)
-		}
-		stdinDone <- err
-	}()
+	stdin := cfg.Streams.Stdin
+
+	if stdin != nil {
+		c.attachReaders.Store(id, stdin)
+
+		go func() {
+			_, err := util.CopyDetachable(conn, stdin, cfg.DetachKeys)
+			if !errors.Is(err, io.ErrClosedPipe) {
+				stdinDone <- err
+			}
+		}()
```

Comment:
sorry, my earlier nit was more saying: why don't we call stdin.Close() here, and avoid the `attachReaders` structure all together?

Reply:
Because we block in CopyDetachable:

```
1 @ 0x43d936 0x44d7bc 0x497791 0x497f45 0x597c27 0x60b137 0x46ea01
#       0x497790        io.(*pipe).read+0xb0                                                                            /nix/store/mf51hzhs63rjzrp8qqzqpj35j140a9bb-go-1.19/share/go/src/io/pipe.go:57
#       0x497f44        io.(*PipeReader).Read+0x24                                                                      /nix/store/mf51hzhs63rjzrp8qqzqpj35j140a9bb-go-1.19/share/go/src/io/pipe.go:136
#       0x597c26        github.com/containers/common/pkg/util.CopyDetachable+0xa6                                       /home/sascha/go/pkg/mod/github.com/containers/common@v0.49.1/pkg/util/copy.go:16
#       0x60b136        github.com/containers/conmon-rs/pkg/client.(*ConmonClient).setupStdioChannels.func2+0x56        /home/sascha/git/conmon-rs/pkg/client/attach.go:205
```